src/languages/translations.ts

```diff
@@ -1,10 +1,50 @@
 import en from './en';
 import es from './es';
 import esES from './es-ES';
+import type {Translation, TranslationBaseValue, TranslationFlatObject} from './types';
+
+/**
+ * Converts an object to it's flattened version.
+ *
+ * Ex:
+ * Input: { common: { yes: "Yes", no: "No" }}
+ * Output: { "common.yes": "Yes", "common.no": "No" }
+ */
+// Necessary to export so that it is accessible to the unit tests
+// eslint-disable-next-line rulesdir/no-inline-named-export
+export function flattenObject(obj: Translation): TranslationFlatObject {
```

Comment:
`TranslationFlatObject` is way to generic (along with `Translation` & `TranslationBaseValue` types). The problem is that we lose type safety using `flattenObject` function:

```ts
const translations = flattenObject(en);
const text2 = translations.nonExistingKey; // TranslationFlatObject, meanwhile it's undefined in reality
```

This is a big problem as we should infer what keys are available to stay type safe when using Localize inside of components. 

What we should do is to flatten the object in Typescript. Something like this ([link](https://stackoverflow.com/questions/58434389/typescript-deep-keyof-of-a-nested-object)):
```ts
type Join<K, P> = K extends string | number ? (P extends string | number ? `${K}${'' extends P ? '' : '.'}${P}` : never) : never;

type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...Array<0>];

type Leaves<T, D extends number = 10> = [D] extends [never] ? never : T extends object ? {[K in keyof T]-?: Join<K, Leaves<T[K], Prev[D]>>}[keyof T] : '';

declare function flattenObject<TTranslation extends Translation>(obj: TTranslation): Record<Leaves<TTranslation>, string | ...>;

const enFlattened = flattenObject2(en);
```

⚠️ These types might be quite slow, so be careful and test before using.

Reply:
However this will be quite irritating to use inside of components as the value is not inferred and it's too generic: `string | string[] | Function`. So we would have to always check the types inside of components 😒

Reply:
Actually I had a similar problem in my side project, something like this could work here but the type is quite complicated ngl 😅 (this would have to be tested as I used it a little bit differentl in my project)

```ts
// Flatten object type
type FlattenObject<T> = CollapseEntries<CreateObjectEntries<T, T>>;

type Entry = {key: string; value: unknown};
type EmptyEntry<T> = {key: ''; value: T};

// Transforms entries to one flattened type
type CollapseEntries<T extends Entry> = {
    [E in T as E['key']]: E['value'];
} extends infer V
    ? {[K in keyof V]: V[K]}
    : never;

type CreateObjectEntries<T, I> = T extends infer U
    ? // Checks that U is an object
      U extends object
        ? {
              // Checks that Key is of type string
              [K in keyof U]-?: K extends string
                  ? // Nested key can be an object, run recursively to the bottom
                    CreateObjectEntries<U[K], I> extends infer E
                      ? E extends Entry
                          ? {
                                key: E['key'] extends '' ? K : `${K}.${E['key']}`;
                                value: E['value'];
                            }
                          : never
                      : never
                  : never;
          }[keyof U] // Builds entry for each key
        : EmptyEntry<U>
    : never;

declare function flattenObject2<TTranslation extends Translation>(obj: TTranslation): FlattenObject<TTranslation>;

const enFlattened = flattenObject2(es);

const type = enFlattened['common.websiteExample']; // string
```

Reply:
Also cc @fabioh8010 as this is quite important to get right 😅